xbmc/guilib/test/TestGUIControlFactory.cpp

```diff
@@ -0,0 +1,665 @@
+                                , public testing::Test
+{
+};
+
+const auto ConditionalVisibilityTests = std::array{
+    ConditionalVisibilityTest{R"(<root>
+                                  <visible>foo</visible>
+                                 </root>)"s,
+                              "foo"s, ""s},
+    ConditionalVisibilityTest{R"(<root>
+                                  <visible allowhiddenfocus="bar">foo</visible>
+                                 </root>)"s,
+                              "foo"s, "bar"s},
+    ConditionalVisibilityTest{R"(<root>
+                                  <visible allowhiddenfocus="bar">foo</visible>
+                                  <visible allowhiddenfocus="foobar">bar</visible>
+                                 </root>)"s,
+                              "[foo] + [bar]"s, "foobar"s},
+    ConditionalVisibilityTest{R"(<root/>)", ""s, ""s, false},
+};
+
+
+struct DimensionTest
+{
+  std::string def;
+  float min, value;
+  bool result = true;
+};
+
+class TestGetDimension : public testing::WithParamInterface<DimensionTest>, public testing::Test
+{
+};
+
+const auto DimensionTests = std::array{
+    DimensionTest{R"(<root><test>0.1</test></root>)"s,
+                  0.f, 0.1f},
+    DimensionTest{R"(<root><test max="3.0">auto</test></root>)"s,
+                  1.f, 3.f},
+    DimensionTest{R"(<root><test max="3.0r">auto</test></root>)"s,
+                  1.f, -2.9f},
+    DimensionTest{R"(<root><test max="3.0%">auto</test></root>)"s,
+                  1.f, 3.f * 0.1f / 100.f},
+    DimensionTest{R"(<root><test min="2.0" max="3.0">auto</test></root>)"s,
+                  2.f, 3.f},
+    DimensionTest{R"(<root><test min="2.0r" max="3.0">auto</test></root>)"s,
+                  -1.9f, 3.f},
+    DimensionTest{R"(<root><test min="2.0%" max="3.0">auto</test></root>)"s,
+                  2.f * 0.1f / 100.f, 3.f},
+    DimensionTest{R"(<root><test>something</test></root>)"s,
+                  0.f, 0.f},
+    DimensionTest{R"(<root><test/></root>)"s,
+                  0.f, 0.f, false},
+    DimensionTest{R"(<root/>)"s,
+                  0.f, 0.f, false},
+};
+
+struct DimensionsTest
+{
+  std::string def;
+  float left, width, min_width;
+  bool result = true;
+};
+
+class TestGetDimensions : public testing::WithParamInterface<DimensionsTest>, public testing::Test
+{
+};
+
+const auto DimensionsTests =  std::array{
+    DimensionsTest{R"(<root><left>0.1</left><width>0.2</width></root>)"s,
+                   0.1f, 0.2f, 0.f},
+    DimensionsTest{R"(<root><left>0.01</left><right>50%</right></root>)"s,
+                   0.01f, 0.1f - 0.1f * 50.f / 100.f - 0.01f, 0.f},
+    DimensionsTest{R"(<root><center_left>0.025f</center_left><right>50%</right></root>)"s,
+                   0.f, (0.05f - 0.025f) * 2, 0.f},
+    DimensionsTest{R"(<root><center_right>0.025f</center_right><width min="0.5" max="0.2">auto</right></root>)"s,
```

Comment:
mismatching elements. is it a <width> element or a <right> element for the second element?

Reply:
fixed.